File: C++/SlidingWindow/239 SlidingWindowMaximum.cpp
// LeetCode 239. Sliding Window Maximum
// Time Complexity: O(n)
// Space Complexity: O(k)

#include <vector>
#include <deque>
using namespace std;

class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        int n = nums.size();
        vector<int> res(n - k + 1);
        deque<int> dq; // stores indices
        int left = 0;

        for (int right = 0; right < n; right++) {

            // Remove indices outside the window
            if (!dq.empty() && dq.front() < right - k + 1) {
                dq.pop_front();
            }

            // Maintain decreasing order
            while (!dq.empty() && nums[dq.back()] <= nums[right]) {
                dq.pop_back();
            }

            dq.push_back(right);

            // Record result once window is valid
            if (right >= k - 1) {
                res[left++] = nums[dq.front()];
            }
        }

        return res;
    }
};

File: Java/All/239 SlidingWindowMaximum.java
// LeetCode 239. Sliding Window Maximum
// Time Complexity: O(n)
// Space Complexity: O(k)

import java.util.ArrayDeque;
import java.util.Deque;

class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
        int n = nums.length;
        int[] res = new int[n - k + 1];
        Deque<Integer> dq = new ArrayDeque<>(); // stores indices
        int left = 0;

        for (int right = 0; right < n; right++) {

            // Remove indices outside the window
            if (!dq.isEmpty() && dq.peekFirst() < right - k + 1) {
                dq.pollFirst();
            }

            // Maintain decreasing order in deque
            while (!dq.isEmpty() && nums[dq.peekLast()] <= nums[right]) {
                dq.pollLast();
            }

            dq.offerLast(right);

            // Record answer once the first window is formed
            if (right >= k - 1) {
                res[left++] = nums[dq.peekFirst()];
            }
        }

        return res;
    }
}
